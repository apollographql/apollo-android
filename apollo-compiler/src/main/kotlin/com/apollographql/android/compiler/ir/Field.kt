package com.apollographql.android.compiler.ir

import com.apollographql.android.compiler.*
import com.squareup.javapoet.*
import java.util.*
import javax.lang.model.element.Modifier

data class Field(
    val responseName: String,
    val fieldName: String,
    val type: String,
    val args: List<Map<String, Any>>? = null,
    val isConditional: Boolean = false,
    val fields: List<Field>? = null,
    val fragmentSpreads: List<String>? = null,
    val inlineFragments: List<InlineFragment>? = null,
    val description: String? = null
) : CodeGenerator {

  override fun toTypeSpec(context: CodeGenerationContext): TypeSpec {
    val fields = if (isNonScalar()) listOf(Field("__typename", "__typename", "String!")) + fields!! else emptyList()
    return SchemaTypeSpecBuilder(
        typeName = formatClassName(),
        fields = fields,
        fragmentSpreads = fragmentSpreads ?: emptyList(),
        inlineFragments = inlineFragments ?: emptyList(),
        context = context
    ).build(Modifier.PUBLIC, Modifier.STATIC)
  }

  fun accessorMethodSpec(context: CodeGenerationContext): MethodSpec {
    return MethodSpec.methodBuilder(responseName.escapeJavaReservedWord())
        .addModifiers(Modifier.PUBLIC)
        .returns(toTypeName(methodResponseType(), context))
        .addStatement("return this.\$L", responseName.escapeJavaReservedWord())
        .let { if (description != null) it.addJavadoc("\$L\n", description) else it }
        .build()
  }

  fun fieldSpec(context: CodeGenerationContext, publicModifier: Boolean = false): FieldSpec {
    return FieldSpec.builder(toTypeName(methodResponseType(), context), responseName.escapeJavaReservedWord())
        .addModifiers(if (publicModifier) Modifier.PUBLIC else Modifier.PRIVATE, Modifier.FINAL)
        .let { if (publicModifier && description != null) it.addJavadoc("\$L\n", description) else it }
        .build()
  }

  fun argumentCodeBlock(): CodeBlock {
    if (args == null || args.isEmpty()) {
      return CodeBlock.builder().add("null").build()
    }
    return jsonMapToCodeBlock(args.fold(HashMap<String, Any>(), { map, arg ->
      map.put(arg["name"].toString(), arg["value"]!!)
      return@fold map
    }))
  }

  private fun jsonMapToCodeBlock(jsonMap: Map<String, Any?>): CodeBlock {
    return jsonMap.entries.map { entry ->
      val codeBuilder = CodeBlock.builder()
      if (entry.value is Map<*, *>) {
        @Suppress("UNCHECKED_CAST")
        codeBuilder.add(".put(\$S, ", entry.key).add("\$L)\n", jsonMapToCodeBlock(entry.value as Map<String, Any?>))
      } else {
        codeBuilder.add(".put(\$S, \$S)\n", entry.key, entry.value).build()
      }
      codeBuilder.build()
    }.fold(CodeBlock.builder().add("new \$T(\$L)\n",
        ClassNames.parameterizedUnmodifiableMapBuilderOf(String::class.java, Any::class.java),
        jsonMap.size
    ).indent(), CodeBlock.Builder::add)
        .unindent()
        .add(".build()").build()
  }

  private fun toTypeName(responseType: String, context: CodeGenerationContext): TypeName {
    val packageName = if (isNonScalar()) "" else context.typesPackage
    return JavaTypeResolver(context, packageName).resolve(responseType, isOptional())
  }

  fun formatClassName() = responseName.capitalize().let { if (isList()) it.singularize() else it }

  private fun methodResponseType(): String {
    if (isNonScalar() || hasFragments()) {
      // For non scalar fields, we use the responseName as the method return type.
      // However, we need to also encode any extra information from the `type` field
      // eg, [lists], nonNulls!, [[nestedLists]], [nonNullLists]!, etc
      val normalizedName = formatClassName()
      if (type.startsWith("[")) {
        // array type
        return if (type.endsWith("!")) "[$normalizedName]!" else "[$normalizedName]"
      } else if (type.endsWith("!")) {
        // non-null type
        return "$normalizedName!"
      } else {
        // nullable type
        return normalizedName
      }
    } else {
      return type
    }
  }

  fun isNonScalar() = hasFragments() || (fields?.any() ?: false)

  fun hasFragments() = (fragmentSpreads?.any() ?: false) || (inlineFragments?.any() ?: false)

  fun isOptional(): Boolean = isConditional || !methodResponseType().endsWith("!")

  fun isList(): Boolean = type.removeSuffix("!").let { it.startsWith('[') && it.endsWith(']') }
}
