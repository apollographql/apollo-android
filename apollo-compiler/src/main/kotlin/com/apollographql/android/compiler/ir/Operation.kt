package com.apollographql.android.compiler.ir

import com.apollographql.android.compiler.SchemaTypeSpecBuilder
import com.apollographql.apollo.api.Operation
import com.squareup.javapoet.TypeSpec
import javax.lang.model.element.Modifier

data class Operation(
    val operationName: String,
    val operationType: String,
    val variables: List<Variable>,
    val source: String,
    val fields: List<Field>,
    val filePath: String,
    val fragmentsReferenced: List<String>
) : CodeGenerator {
  override fun toTypeSpec(context: CodeGenerationContext): TypeSpec =
      SchemaTypeSpecBuilder(DATA_TYPE_NAME, fields, emptyList(), emptyList(), context)
          .build(Modifier.PUBLIC, Modifier.STATIC)
          .toBuilder()
          .addSuperinterface(Operation.Data::class.java)
          .build()

  companion object {
    val DATA_TYPE_NAME = "Data"
  }
}
