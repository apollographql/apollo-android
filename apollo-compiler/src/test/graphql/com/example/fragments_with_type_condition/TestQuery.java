package com.example.fragments_with_type_condition;

import com.apollographql.apollo.api.Field;
import com.apollographql.apollo.api.FragmentResponseFieldMapper;
import com.apollographql.apollo.api.Operation;
import com.apollographql.apollo.api.OperationName;
import com.apollographql.apollo.api.Query;
import com.apollographql.apollo.api.ResponseFieldMapper;
import com.apollographql.apollo.api.ResponseReader;
import com.apollographql.apollo.api.internal.Optional;
import com.example.fragments_with_type_condition.fragment.DroidDetails;
import com.example.fragments_with_type_condition.fragment.HumanDetails;
import java.io.IOException;
import java.lang.Object;
import java.lang.Override;
import java.lang.String;
import javax.annotation.Generated;
import javax.annotation.Nonnull;
import javax.annotation.Nullable;

@Generated("Apollo GraphQL")
public final class TestQuery implements Query<TestQuery.Data, Optional<TestQuery.Data>, Operation.Variables> {
  public static final String OPERATION_DEFINITION = "query TestQuery {\n"
      + "  r2: hero {\n"
      + "    __typename\n"
      + "    ...HumanDetails\n"
      + "    ...DroidDetails\n"
      + "  }\n"
      + "  luke: hero {\n"
      + "    __typename\n"
      + "    ...HumanDetails\n"
      + "    ...DroidDetails\n"
      + "  }\n"
      + "}";

  public static final String QUERY_DOCUMENT = OPERATION_DEFINITION + "\n"
   + HumanDetails.FRAGMENT_DEFINITION + "\n"
   + DroidDetails.FRAGMENT_DEFINITION;

  private static final OperationName OPERATION_NAME = new OperationName() {
    @Override
    public String name() {
      return "TestQuery";
    }
  };

  private final Operation.Variables variables;

  public TestQuery() {
    this.variables = Operation.EMPTY_VARIABLES;
  }

  @Override
  public String queryDocument() {
    return QUERY_DOCUMENT;
  }

  @Override
  public Optional<TestQuery.Data> wrapData(TestQuery.Data data) {
    return Optional.fromNullable(data);
  }

  @Override
  public Operation.Variables variables() {
    return variables;
  }

  @Override
  public ResponseFieldMapper<TestQuery.Data> responseFieldMapper() {
    return new Data.Mapper();
  }

  public static Builder builder() {
    return new Builder();
  }

  @Override
  public OperationName name() {
    return OPERATION_NAME;
  }

  public static final class Builder {
    Builder() {
    }

    public TestQuery build() {
      return new TestQuery();
    }
  }

  public static class Data implements Operation.Data {
    private final Optional<R2> r2;

    private final Optional<Luke> luke;

    private volatile String $toString;

    private volatile int $hashCode;

    private volatile boolean $hashCodeMemoized;

    public Data(@Nullable R2 r2, @Nullable Luke luke) {
      this.r2 = Optional.fromNullable(r2);
      this.luke = Optional.fromNullable(luke);
    }

    public Optional<R2> r2() {
      return this.r2;
    }

    public Optional<Luke> luke() {
      return this.luke;
    }

    @Override
    public String toString() {
      if ($toString == null) {
        $toString = "Data{"
          + "r2=" + r2 + ", "
          + "luke=" + luke
          + "}";
      }
      return $toString;
    }

    @Override
    public boolean equals(Object o) {
      if (o == this) {
        return true;
      }
      if (o instanceof Data) {
        Data that = (Data) o;
        return this.r2.equals(that.r2)
         && this.luke.equals(that.luke);
      }
      return false;
    }

    @Override
    public int hashCode() {
      if (!$hashCodeMemoized) {
        int h = 1;
        h *= 1000003;
        h ^= r2.hashCode();
        h *= 1000003;
        h ^= luke.hashCode();
        $hashCode = h;
        $hashCodeMemoized = true;
      }
      return $hashCode;
    }

    public static final class Mapper implements ResponseFieldMapper<Data> {
      final R2.Mapper r2FieldMapper = new R2.Mapper();

      final Luke.Mapper lukeFieldMapper = new Luke.Mapper();

      final Field[] fields = {
        Field.forObject("r2", "hero", null, true, new Field.ObjectReader<R2>() {
          @Override public R2 read(final ResponseReader reader) throws IOException {
            return r2FieldMapper.map(reader);
          }
        }),
        Field.forObject("luke", "hero", null, true, new Field.ObjectReader<Luke>() {
          @Override public Luke read(final ResponseReader reader) throws IOException {
            return lukeFieldMapper.map(reader);
          }
        })
      };

      @Override
      public Data map(ResponseReader reader) throws IOException {
        final R2 r2 = reader.read(fields[0]);
        final Luke luke = reader.read(fields[1]);
        return new Data(r2, luke);
      }
    }
  }

  public static class R2 {
    private final @Nonnull String __typename;

    private final @Nonnull Fragments fragments;

    private volatile String $toString;

    private volatile int $hashCode;

    private volatile boolean $hashCodeMemoized;

    public R2(@Nonnull String __typename, @Nonnull Fragments fragments) {
      this.__typename = __typename;
      this.fragments = fragments;
    }

    public @Nonnull String __typename() {
      return this.__typename;
    }

    public @Nonnull Fragments fragments() {
      return this.fragments;
    }

    @Override
    public String toString() {
      if ($toString == null) {
        $toString = "R2{"
          + "__typename=" + __typename + ", "
          + "fragments=" + fragments
          + "}";
      }
      return $toString;
    }

    @Override
    public boolean equals(Object o) {
      if (o == this) {
        return true;
      }
      if (o instanceof R2) {
        R2 that = (R2) o;
        return this.__typename.equals(that.__typename)
         && this.fragments.equals(that.fragments);
      }
      return false;
    }

    @Override
    public int hashCode() {
      if (!$hashCodeMemoized) {
        int h = 1;
        h *= 1000003;
        h ^= __typename.hashCode();
        h *= 1000003;
        h ^= fragments.hashCode();
        $hashCode = h;
        $hashCodeMemoized = true;
      }
      return $hashCode;
    }

    public static class Fragments {
      private final @Nonnull HumanDetails humanDetails;

      private final @Nonnull DroidDetails droidDetails;

      private volatile String $toString;

      private volatile int $hashCode;

      private volatile boolean $hashCodeMemoized;

      public Fragments(@Nonnull HumanDetails humanDetails, @Nonnull DroidDetails droidDetails) {
        this.humanDetails = humanDetails;
        this.droidDetails = droidDetails;
      }

      public @Nonnull HumanDetails humanDetails() {
        return this.humanDetails;
      }

      public @Nonnull DroidDetails droidDetails() {
        return this.droidDetails;
      }

      @Override
      public String toString() {
        if ($toString == null) {
          $toString = "Fragments{"
            + "humanDetails=" + humanDetails + ", "
            + "droidDetails=" + droidDetails
            + "}";
        }
        return $toString;
      }

      @Override
      public boolean equals(Object o) {
        if (o == this) {
          return true;
        }
        if (o instanceof Fragments) {
          Fragments that = (Fragments) o;
          return this.humanDetails.equals(that.humanDetails)
           && this.droidDetails.equals(that.droidDetails);
        }
        return false;
      }

      @Override
      public int hashCode() {
        if (!$hashCodeMemoized) {
          int h = 1;
          h *= 1000003;
          h ^= humanDetails.hashCode();
          h *= 1000003;
          h ^= droidDetails.hashCode();
          $hashCode = h;
          $hashCodeMemoized = true;
        }
        return $hashCode;
      }

      public static final class Mapper implements FragmentResponseFieldMapper<Fragments> {
        final HumanDetails.Mapper humanDetailsFieldMapper = new HumanDetails.Mapper();

        final DroidDetails.Mapper droidDetailsFieldMapper = new DroidDetails.Mapper();

        @Override
        public @Nonnull Fragments map(ResponseReader reader, @Nonnull String conditionalType) throws
            IOException {
          HumanDetails humanDetails = null;
          DroidDetails droidDetails = null;
          if (HumanDetails.POSSIBLE_TYPES.contains(conditionalType)) {
            humanDetails = humanDetailsFieldMapper.map(reader);
          }
          if (DroidDetails.POSSIBLE_TYPES.contains(conditionalType)) {
            droidDetails = droidDetailsFieldMapper.map(reader);
          }
          return new Fragments(humanDetails, droidDetails);
        }
      }
    }

    public static final class Mapper implements ResponseFieldMapper<R2> {
      final Fragments.Mapper fragmentsFieldMapper = new Fragments.Mapper();

      final Field[] fields = {
        Field.forString("__typename", "__typename", null, false),
        Field.forConditionalType("__typename", "__typename", new Field.ConditionalTypeReader<Fragments>() {
          @Override
          public Fragments read(String conditionalType, ResponseReader reader) throws IOException {
            return fragmentsFieldMapper.map(reader, conditionalType);
          }
        })
      };

      @Override
      public R2 map(ResponseReader reader) throws IOException {
        final String __typename = reader.read(fields[0]);
        final Fragments fragments = reader.read(fields[1]);
        return new R2(__typename, fragments);
      }
    }
  }

  public static class Luke {
    private final @Nonnull String __typename;

    private final @Nonnull Fragments fragments;

    private volatile String $toString;

    private volatile int $hashCode;

    private volatile boolean $hashCodeMemoized;

    public Luke(@Nonnull String __typename, @Nonnull Fragments fragments) {
      this.__typename = __typename;
      this.fragments = fragments;
    }

    public @Nonnull String __typename() {
      return this.__typename;
    }

    public @Nonnull Fragments fragments() {
      return this.fragments;
    }

    @Override
    public String toString() {
      if ($toString == null) {
        $toString = "Luke{"
          + "__typename=" + __typename + ", "
          + "fragments=" + fragments
          + "}";
      }
      return $toString;
    }

    @Override
    public boolean equals(Object o) {
      if (o == this) {
        return true;
      }
      if (o instanceof Luke) {
        Luke that = (Luke) o;
        return this.__typename.equals(that.__typename)
         && this.fragments.equals(that.fragments);
      }
      return false;
    }

    @Override
    public int hashCode() {
      if (!$hashCodeMemoized) {
        int h = 1;
        h *= 1000003;
        h ^= __typename.hashCode();
        h *= 1000003;
        h ^= fragments.hashCode();
        $hashCode = h;
        $hashCodeMemoized = true;
      }
      return $hashCode;
    }

    public static class Fragments {
      private final @Nonnull HumanDetails humanDetails;

      private final @Nonnull DroidDetails droidDetails;

      private volatile String $toString;

      private volatile int $hashCode;

      private volatile boolean $hashCodeMemoized;

      public Fragments(@Nonnull HumanDetails humanDetails, @Nonnull DroidDetails droidDetails) {
        this.humanDetails = humanDetails;
        this.droidDetails = droidDetails;
      }

      public @Nonnull HumanDetails humanDetails() {
        return this.humanDetails;
      }

      public @Nonnull DroidDetails droidDetails() {
        return this.droidDetails;
      }

      @Override
      public String toString() {
        if ($toString == null) {
          $toString = "Fragments{"
            + "humanDetails=" + humanDetails + ", "
            + "droidDetails=" + droidDetails
            + "}";
        }
        return $toString;
      }

      @Override
      public boolean equals(Object o) {
        if (o == this) {
          return true;
        }
        if (o instanceof Fragments) {
          Fragments that = (Fragments) o;
          return this.humanDetails.equals(that.humanDetails)
           && this.droidDetails.equals(that.droidDetails);
        }
        return false;
      }

      @Override
      public int hashCode() {
        if (!$hashCodeMemoized) {
          int h = 1;
          h *= 1000003;
          h ^= humanDetails.hashCode();
          h *= 1000003;
          h ^= droidDetails.hashCode();
          $hashCode = h;
          $hashCodeMemoized = true;
        }
        return $hashCode;
      }

      public static final class Mapper implements FragmentResponseFieldMapper<Fragments> {
        final HumanDetails.Mapper humanDetailsFieldMapper = new HumanDetails.Mapper();

        final DroidDetails.Mapper droidDetailsFieldMapper = new DroidDetails.Mapper();

        @Override
        public @Nonnull Fragments map(ResponseReader reader, @Nonnull String conditionalType) throws
            IOException {
          HumanDetails humanDetails = null;
          DroidDetails droidDetails = null;
          if (HumanDetails.POSSIBLE_TYPES.contains(conditionalType)) {
            humanDetails = humanDetailsFieldMapper.map(reader);
          }
          if (DroidDetails.POSSIBLE_TYPES.contains(conditionalType)) {
            droidDetails = droidDetailsFieldMapper.map(reader);
          }
          return new Fragments(humanDetails, droidDetails);
        }
      }
    }

    public static final class Mapper implements ResponseFieldMapper<Luke> {
      final Fragments.Mapper fragmentsFieldMapper = new Fragments.Mapper();

      final Field[] fields = {
        Field.forString("__typename", "__typename", null, false),
        Field.forConditionalType("__typename", "__typename", new Field.ConditionalTypeReader<Fragments>() {
          @Override
          public Fragments read(String conditionalType, ResponseReader reader) throws IOException {
            return fragmentsFieldMapper.map(reader, conditionalType);
          }
        })
      };

      @Override
      public Luke map(ResponseReader reader) throws IOException {
        final String __typename = reader.read(fields[0]);
        final Fragments fragments = reader.read(fields[1]);
        return new Luke(__typename, fragments);
      }
    }
  }
}
