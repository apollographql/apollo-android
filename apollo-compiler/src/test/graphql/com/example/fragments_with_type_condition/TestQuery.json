{
	"operations": [
		{
			"filePath": "src/test/graphql/com/example/fragments_with_type_condition/TestQuery.graphql",
			"operationName": "TestQuery",
			"operationType": "query",
			"variables": [],
			"source": "query TestQuery {\n  r2: hero {\n    __typename\n    ...HumanDetails\n    ...DroidDetails\n  }\n  luke: hero {\n    __typename\n    ...HumanDetails\n    ...DroidDetails\n  }\n}",
			"fields": [
				{
					"responseName": "r2",
					"fieldName": "hero",
					"type": "Character",
					"fields": [],
					"fragmentSpreads": [
						"HumanDetails",
						"DroidDetails"
					],
					"inlineFragments": []
				},
				{
					"responseName": "luke",
					"fieldName": "hero",
					"type": "Character",
					"fields": [],
					"fragmentSpreads": [
						"HumanDetails",
						"DroidDetails"
					],
					"inlineFragments": []
				}
			],
			"fragmentsReferenced": [
				"HumanDetails",
				"DroidDetails"
			]
		}
	],
	"fragments": [
		{
			"filePath": "src/test/graphql/com/example/fragments_with_type_condition/TestQuery.graphql",
			"fragmentName": "HumanDetails",
			"source": "fragment HumanDetails on Human {\n  __typename\n  name\n  height\n}",
			"typeCondition": "Human",
			"possibleTypes": [
				"Human"
			],
			"fields": [
				{
					"responseName": "name",
					"fieldName": "name",
					"type": "String!",
					"description": "What this human calls themselves"
				},
				{
					"responseName": "height",
					"fieldName": "height",
					"type": "Float",
					"description": "Height in the preferred unit, default is meters"
				}
			],
			"fragmentSpreads": [],
			"inlineFragments": [],
			"fragmentsReferenced": []
		},
		{
			"filePath": "src/test/graphql/com/example/fragments_with_type_condition/TestQuery.graphql",
			"fragmentName": "DroidDetails",
			"source": "fragment DroidDetails on Droid {\n  __typename\n  name\n  primaryFunction\n}",
			"typeCondition": "Droid",
			"possibleTypes": [
				"Droid"
			],
			"fields": [
				{
					"responseName": "name",
					"fieldName": "name",
					"type": "String!",
					"description": "What others call this droid"
				},
				{
					"responseName": "primaryFunction",
					"fieldName": "primaryFunction",
					"type": "String",
					"description": "This droid's primary function"
				}
			],
			"fragmentSpreads": [],
			"inlineFragments": [],
			"fragmentsReferenced": []
		}
	],
	"typesUsed": []
}