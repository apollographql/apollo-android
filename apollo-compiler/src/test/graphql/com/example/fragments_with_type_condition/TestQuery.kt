// AUTO-GENERATED FILE. DO NOT MODIFY.
//
// This class was automatically generated by Apollo GraphQL plugin from the GraphQL queries it found.
// It should not be modified by hand.
//
package com.example.fragments_with_type_condition

import com.apollographql.apollo.api.CustomScalarAdapters
import com.apollographql.apollo.api.CustomScalarAdapters.Companion.DEFAULT
import com.apollographql.apollo.api.Operation
import com.apollographql.apollo.api.OperationName
import com.apollographql.apollo.api.Query
import com.apollographql.apollo.api.Response
import com.apollographql.apollo.api.internal.OperationRequestBodyComposer
import com.apollographql.apollo.api.internal.QueryDocumentMinifier
import com.apollographql.apollo.api.internal.ResponseFieldMapper
import com.apollographql.apollo.api.internal.ResponseFieldMarshaller
import com.apollographql.apollo.api.internal.SimpleOperationResponseParser
import com.apollographql.apollo.api.internal.Throws
import com.example.fragments_with_type_condition.adapter.TestQuery_ResponseAdapter
import com.example.fragments_with_type_condition.fragment.DroidDetail
import com.example.fragments_with_type_condition.fragment.HumanDetail
import kotlin.Boolean
import kotlin.Double
import kotlin.String
import kotlin.Suppress
import okio.Buffer
import okio.BufferedSource
import okio.ByteString
import okio.IOException

@Suppress("NAME_SHADOWING", "UNUSED_ANONYMOUS_PARAMETER", "LocalVariableName",
    "RemoveExplicitTypeArguments", "NestedLambdaShadowedImplicitParameter", "PropertyName",
    "RemoveRedundantQualifierName")
class TestQuery : Query<TestQuery.Data, Operation.Variables> {
  override fun operationId(): String = OPERATION_ID

  override fun queryDocument(): String = QUERY_DOCUMENT

  override fun variables(): Operation.Variables = Operation.EMPTY_VARIABLES

  override fun name(): OperationName = OPERATION_NAME

  override fun responseFieldMapper(): ResponseFieldMapper<Data> {
    return ResponseFieldMapper { reader ->
      TestQuery_ResponseAdapter.fromResponse(reader)
    }
  }

  @Throws(IOException::class)
  override fun parse(source: BufferedSource, customScalarAdapters: CustomScalarAdapters):
      Response<Data> {
    return SimpleOperationResponseParser.parse(source, this, customScalarAdapters)
  }

  @Throws(IOException::class)
  override fun parse(byteString: ByteString, customScalarAdapters: CustomScalarAdapters):
      Response<Data> {
    return parse(Buffer().write(byteString), customScalarAdapters)
  }

  @Throws(IOException::class)
  override fun parse(source: BufferedSource): Response<Data> {
    return parse(source, DEFAULT)
  }

  @Throws(IOException::class)
  override fun parse(byteString: ByteString): Response<Data> {
    return parse(byteString, DEFAULT)
  }

  override fun composeRequestBody(customScalarAdapters: CustomScalarAdapters): ByteString {
    return OperationRequestBodyComposer.compose(
      operation = this,
      autoPersistQueries = false,
      withQueryDocument = true,
      customScalarAdapters = customScalarAdapters
    )
  }

  override fun composeRequestBody(): ByteString = OperationRequestBodyComposer.compose(
    operation = this,
    autoPersistQueries = false,
    withQueryDocument = true,
    customScalarAdapters = DEFAULT
  )

  override fun composeRequestBody(
    autoPersistQueries: Boolean,
    withQueryDocument: Boolean,
    customScalarAdapters: CustomScalarAdapters
  ): ByteString = OperationRequestBodyComposer.compose(
    operation = this,
    autoPersistQueries = autoPersistQueries,
    withQueryDocument = withQueryDocument,
    customScalarAdapters = customScalarAdapters
  )

  /**
   * The query type, represents all of the entry points into our object graph
   */
  data class Data(
    val r2: R2?,
    val luke: Luke?
  ) : Operation.Data {
    override fun marshaller(): ResponseFieldMarshaller {
      return ResponseFieldMarshaller { writer ->
        TestQuery_ResponseAdapter.Data.toResponse(writer, this)
      }
    }

    /**
     * A character from the Star Wars universe
     */
    interface R2 {
      val __typename: String

      fun marshaller(): ResponseFieldMarshaller

      interface Human : R2, HumanDetail {
        override val __typename: String

        /**
         * What this human calls themselves
         */
        override val name: String

        /**
         * Height in the preferred unit, default is meters
         */
        override val height: Double?

        override fun marshaller(): ResponseFieldMarshaller
      }

      interface Droid : R2, DroidDetail {
        override val __typename: String

        /**
         * What others call this droid
         */
        override val name: String

        /**
         * This droid's primary function
         */
        override val primaryFunction: String?

        override fun marshaller(): ResponseFieldMarshaller
      }

      data class HumanR2(
        override val __typename: String,
        /**
         * What this human calls themselves
         */
        override val name: String,
        /**
         * Height in the preferred unit, default is meters
         */
        override val height: Double?
      ) : R2, Human, HumanDetail {
        override fun marshaller(): ResponseFieldMarshaller {
          return ResponseFieldMarshaller { writer ->
            TestQuery_ResponseAdapter.Data.R2.HumanR2.toResponse(writer, this)
          }
        }
      }

      data class DroidR2(
        override val __typename: String,
        /**
         * What others call this droid
         */
        override val name: String,
        /**
         * This droid's primary function
         */
        override val primaryFunction: String?
      ) : R2, Droid, DroidDetail {
        override fun marshaller(): ResponseFieldMarshaller {
          return ResponseFieldMarshaller { writer ->
            TestQuery_ResponseAdapter.Data.R2.DroidR2.toResponse(writer, this)
          }
        }
      }

      data class OtherR2(
        override val __typename: String
      ) : R2 {
        override fun marshaller(): ResponseFieldMarshaller {
          return ResponseFieldMarshaller { writer ->
            TestQuery_ResponseAdapter.Data.R2.OtherR2.toResponse(writer, this)
          }
        }
      }

      companion object {
        fun R2.asHuman(): Human? = this as? Human

        fun R2.humanDetails(): HumanDetail? = this as? HumanDetail

        fun R2.asDroid(): Droid? = this as? Droid

        fun R2.droidDetails(): DroidDetail? = this as? DroidDetail
      }
    }

    /**
     * A character from the Star Wars universe
     */
    interface Luke {
      val __typename: String

      fun marshaller(): ResponseFieldMarshaller

      interface Human : Luke, HumanDetail {
        override val __typename: String

        /**
         * What this human calls themselves
         */
        override val name: String

        /**
         * Height in the preferred unit, default is meters
         */
        override val height: Double?

        override fun marshaller(): ResponseFieldMarshaller
      }

      interface Droid : Luke, DroidDetail {
        override val __typename: String

        /**
         * What others call this droid
         */
        override val name: String

        /**
         * This droid's primary function
         */
        override val primaryFunction: String?

        override fun marshaller(): ResponseFieldMarshaller
      }

      data class HumanLuke(
        override val __typename: String,
        /**
         * What this human calls themselves
         */
        override val name: String,
        /**
         * Height in the preferred unit, default is meters
         */
        override val height: Double?
      ) : Luke, Human, HumanDetail {
        override fun marshaller(): ResponseFieldMarshaller {
          return ResponseFieldMarshaller { writer ->
            TestQuery_ResponseAdapter.Data.Luke.HumanLuke.toResponse(writer, this)
          }
        }
      }

      data class DroidLuke(
        override val __typename: String,
        /**
         * What others call this droid
         */
        override val name: String,
        /**
         * This droid's primary function
         */
        override val primaryFunction: String?
      ) : Luke, Droid, DroidDetail {
        override fun marshaller(): ResponseFieldMarshaller {
          return ResponseFieldMarshaller { writer ->
            TestQuery_ResponseAdapter.Data.Luke.DroidLuke.toResponse(writer, this)
          }
        }
      }

      data class OtherLuke(
        override val __typename: String
      ) : Luke {
        override fun marshaller(): ResponseFieldMarshaller {
          return ResponseFieldMarshaller { writer ->
            TestQuery_ResponseAdapter.Data.Luke.OtherLuke.toResponse(writer, this)
          }
        }
      }

      companion object {
        fun Luke.asHuman(): Human? = this as? Human

        fun Luke.humanDetails(): HumanDetail? = this as? HumanDetail

        fun Luke.asDroid(): Droid? = this as? Droid

        fun Luke.droidDetails(): DroidDetail? = this as? DroidDetail
      }
    }
  }

  companion object {
    const val OPERATION_ID: String =
        "919cec7210259fa24fc6026fe680b96f357c14ebf3c8a734979dcfb819685d6a"

    val QUERY_DOCUMENT: String = QueryDocumentMinifier.minify(
          """
          |query TestQuery {
          |  r2: hero {
          |    __typename
          |    ...HumanDetails
          |    ...DroidDetails
          |  }
          |  luke: hero {
          |    __typename
          |    ...HumanDetails
          |    ...DroidDetails
          |  }
          |}
          |fragment HumanDetails on Human {
          |  __typename
          |  name
          |  height
          |}
          |fragment DroidDetails on Droid {
          |  __typename
          |  name
          |  primaryFunction
          |}
          """.trimMargin()
        )

    val OPERATION_NAME: OperationName = object : OperationName {
      override fun name(): String {
        return "TestQuery"
      }
    }
  }
}
