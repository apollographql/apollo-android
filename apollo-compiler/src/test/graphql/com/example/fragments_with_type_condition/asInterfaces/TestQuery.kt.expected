//
// AUTO-GENERATED FILE. DO NOT MODIFY.
//
// This class was automatically generated by Apollo GraphQL version '3.0.0-dev7-SNAPSHOT'.
//
package com.example.fragments_with_type_condition

import com.apollographql.apollo3.api.Query
import com.apollographql.apollo3.api.ResponseAdapter
import com.apollographql.apollo3.api.ResponseAdapterCache
import com.apollographql.apollo3.api.ResponseField
import com.apollographql.apollo3.api.json.JsonWriter
import com.apollographql.apollo3.api.obj
import com.example.fragments_with_type_condition.adapter.TestQuery_ResponseAdapter
import com.example.fragments_with_type_condition.fragment.HumanDetails
import com.example.fragments_with_type_condition.responsefields.TestQuery_ResponseFields
import kotlin.Double
import kotlin.String
import kotlin.collections.List

class TestQuery : Query<TestQuery.Data> {
  override fun id(): String = OPERATION_ID

  override fun document(): String = OPERATION_DOCUMENT

  override fun name(): String = OPERATION_NAME

  override fun serializeVariables(writer: JsonWriter, responseAdapterCache: ResponseAdapterCache) {
    // This operation doesn't have variables
  }

  override fun adapter(): ResponseAdapter<Data> {
    return TestQuery_ResponseAdapter.Data.obj()
  }

  override fun responseFields(): List<ResponseField.FieldSet> =
      TestQuery_ResponseFields.fields.first().fieldSets

  data class Data(
    val r2: R2?,
    val luke: Luke?
  ) : Query.Data {
    interface R2 {
      val __typename: String

      companion object {
        fun R2.HumanDetails() = this as? HumanDetails

        fun R2.asCharacterHuman() = this as? HumanR2

        fun R2.asCharacterDroid() = this as? DroidR2
      }
    }

    data class HumanR2(
      override val __typename: String,
      /**
       * What this human calls themselves
       */
      override val name: String,
      /**
       * Height in the preferred unit, default is meters
       */
      override val height: Double?
    ) : R2, HumanDetails

    data class DroidR2(
      override val __typename: String,
      /**
       * What others call this droid
       */
      val name: String,
      /**
       * This droid's primary function
       */
      val primaryFunction: String?
    ) : R2

    data class OtherR2(
      override val __typename: String
    ) : R2

    interface Luke {
      val __typename: String

      companion object {
        fun Luke.HumanDetails() = this as? HumanDetails

        fun Luke.asCharacterHuman() = this as? HumanLuke

        fun Luke.asCharacterDroid() = this as? DroidLuke
      }
    }

    data class HumanLuke(
      override val __typename: String,
      /**
       * What this human calls themselves
       */
      override val name: String,
      /**
       * Height in the preferred unit, default is meters
       */
      override val height: Double?
    ) : Luke, HumanDetails

    data class DroidLuke(
      override val __typename: String,
      /**
       * What others call this droid
       */
      val name: String,
      /**
       * This droid's primary function
       */
      val primaryFunction: String?
    ) : Luke

    data class OtherLuke(
      override val __typename: String
    ) : Luke
  }

  companion object {
    const val OPERATION_ID: String =
        "489a08c53439e4fcdbdf6ec743933e2709d52bb538e2c5fda5d7656dd2c2491f"

    /**
     * The minimized GraphQL document being sent to the server to save a few bytes.
     * The un-minimized version is:
     *
     * query TestQuery {
     *   r2: hero {
     *     __typename
     *     ...HumanDetails
     *     ...DroidDetails
     *   }
     *   luke: hero {
     *     __typename
     *     ...HumanDetails
     *     ...DroidDetails
     *   }
     * }
     */
    const val OPERATION_DOCUMENT: String =
        "query TestQuery { r2: hero { __typename ...HumanDetails ...DroidDetails } luke: hero { __typename ...HumanDetails ...DroidDetails } }"

    const val OPERATION_NAME: String = "TestQuery"
  }
}
